### ISSUE TITLE
Sparse matrices support

### ISSUE THREAD

**Issue by raver119** [eclipse/deeplearning4j#4990]
We should provide full support for sparse NDArrays

**Comment by Lundez** [comment]
Any updates on sparse matrices?

**Comment by agibsonccc** [comment]
@Lundez we had a half done implementation that was removed due to it being better to not have the feature rather than a half done one that's confusing to users. Open a more specific feature request issue and maybe we can do something for you, but for now, every op supporting sparse efficiently isn't really something that's had a ton of demand.

There's certain use cases where it's really important, but the ROI wasn't quite there to fully implement and support sparse. If you would like to contribute, you are welcome to.

**Comment by Lundez** [comment]
@agibsonccc I see.
It's mainly for feature vectors that are incredibly sparse.
Like BagOfWords, TfIdf & more which would save a ton of memory by using sparse vectors. Especially if you apply TfIdf with a windowed size like all unigrams & bigrams.
Simple things like `multiplication`, `addition` and more would be good, but solvers like `QR` / `SVD` could be applied by transforming it into a dense matrix.

If there's a lot of work do you've any estimate of time required?

**Comment by agibsonccc** [comment]
@Lundez yes of course we're aware of that and attempted that before.  The ROI just isn't there for that. We have hundreds of operations that would effectively have to become sparse aware. It's not just flipping a switch. It's actually a ton of work we're not really looking at doing at this point. 

We tried compromising by only supporting the various sparse implementations, but even the sparse blas story isn't as clear. Implementing it part way might work for you, but will confuse a bunch of other people who would expect sparse (and for that matter, complex types) to "just work" with any op.

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
